python/tank/util/metrics.py

```diff
@@ -0,0 +1,391 @@
+                metrics = [self._queue.popleft() for i in range(0, count)]
+            except Exception:
+                pass
+
+        self._lock.release()
+
+        return metrics
+
+
+class MetricsDispatcher(object):
+    """This class manages 1 or more worker threads dispatching toolkit metrics.
+
+    After initializing the object, the `start()` method is called to
+    spin up worker threads for dispatching logged metrics. The `stop()` method
+    is later called to stop the worker threads.
+
+    """
+
+    def __init__(self, engine, tk, num_workers=1):
+        """Initialize the dispatcher object.
+
+        :param engine: Engine instance, used for logging
+        :param tk: Toolkit api instance
+        :param workers: The number of worker threads to start.
+
+        """
+
+        self._engine = engine
+        self._tk = tk
+        self._num_workers = num_workers
+        self._workers = []
+        self._dispatching = False
+
+    def start(self):
+        """Starts up the workers for dispatching logged metrics.
+
+        If called on an already dispatching instance, then result is a no-op.
+
+        """
+
+        if self._dispatching:
+            self._engine.log_debug(
+                "Metrics dispatching already started. Doing nothing.")
+            return
+
+        # if metrics are not supported, then no reason to process the queue
+        if not self._metrics_supported(self._tk.shotgun):
+            self._engine.log_debug(
+                "Metrics not supported for this version of Shotgun.")
+            return
+
+        # start the dispatch workers to use this queue
+        for i in range(self._num_workers):
+            worker = _MetricsDispatchWorkerThread(self._tk, self._engine)
+            worker.start()
+            self._engine.log_debug("Added worker thread: %s" % (worker,))
+            self._workers.append(worker)
+
+        self._dispatching = True
+
+    def stop(self):
+        """Instructs all worker threads to stop processing metrics."""
+        for worker in self.workers:
+            worker.halt()
+
+        self._dispatching = False
+        self._workers = []
+
+    @property
+    def dispatching(self):
+        """True if started and dispatching metrics."""
+        return self._dispatching
+
+    @property
+    def workers(self):
+        """A list of workers threads dispatching metrics from the queue."""
+        return self._workers
+
+    def _metrics_supported(self, sg_connection):
+        """Returns True if server supports the metrics api endpoint."""
+
+        # TODO: update the version number once the endpoint is available
+        self._metrics_ok = (
+            sg_connection.server_caps.version and
+            sg_connection.server_caps.version >= (7, 0, 0)
+        )
+
+        # TODO: remove after testing
+        self._metrics_ok = True
+        return self._metrics_ok
+
+
+class _MetricsDispatchWorkerThread(Thread):
```

Comment:
Thought: I think we should drop the underscore. We have 'private' classes in other places and we don't generally use that sort of convention. Does that make sense?